### ISSUE TITLE
EMIRGE Amplicon: n_alignments not intialized

### ISSUE THREAD

**Issue by TheOneHyer** [csmiller/EMIRGE#19]
Upon running emirge_amplicon.py assembly with the following command:

emirge_amplicon.py assembly --phred33 --iterations 120 --fasta_db /srv/databases/emirge/SSURef_111_candidate_db.fasta --bowtie_db /srv/databases/emirge/SSU_111_candidate_db --mapping TL04arch.mapped.sorted.bam --max_read_length 126 --insert_mean 100 --insert_stddev 10 -1 ...TL02arch-Overland-DNA2.forward.decontam.adapt_trim.qual_trim.fastq.gz -2 ...TL02arch-Overland-DNA2.reverse.decontam.adapt_trim.qual_trim.fastq

the program crashes with:

EMIRGE started at Fri Feb 19 17:02:01 2016
Rewriting reads with indices in headers at Fri Feb 19 17:02:01 2016...
DONE Rewriting reads with indexes in headers at Fri Feb 19 17:03:03 2016 [0:01:02.031644]...
Number of reads (or read pairs) in input file(s): 5806520
Preallocating reads and quals in memory at Fri Feb 19 17:03:03 2016...
DONE Preallocating reads and quals in memory at Fri Feb 19 17:03:47 2016 [0:00:44.743259]...
Beginning initialization at Fri Feb 19 17:03:47 2016...
Reading bam file ...TL04arch.mapped.sorted.bam at Fri Feb 19 17:03:47 2016...
Traceback (most recent call last):
  File "/bin/emirge_amplicon.py", line 1576, in <module>
    main()
  File "/bin/emirge_amplicon.py", line 1565, in main
    em.initialize_EM(options.mapping, options.fasta_db, randomize_priors = options.randomize_init_priors)
  File "/bin/emirge_amplicon.py", line 384, in initialize_EM
    self.read_bam(bam_filename, reference_fasta_filename)
  File "/bin/emirge_amplicon.py", line 334, in read_bam
    _emirge.process_bamfile(self, BOWTIE_ASCII_OFFSET)
  File "_emirge_amplicon.pyx", line 553, in _emirge_amplicon.process_bamfile (_emirge_amplicon.c:7241)
**AttributeError: 'EM' object has no attribute 'n_alignments'**

I've been investigating the error and this is what seems to be happening:

emirge_amplicon.py

line 1576 calls main()
line 1533 initializes the em class, there is not a self.n_alignments variable initialized with the class
line 1565 calls em.initialize_EM
line 384 in initialize_EM(args) calls self.read_bam(bam_filename, reference_fasta_filename)
line 334 in read_bam calls _emirge.process_bamfile(self, BOWTIE_ASCII_OFFSET)
\# The prior command leads to a function in _emirge_amplicon.pyx called process_bamfile
line 553 in the aforementioned function calls bamfile_data = np.empty((em.n_alignments, 6), dtype=np.uint32)
\# Note: self is passed to process_bamfile which is now called em
\# This is where the error is thrown.

n_alignments appears to be generated by get_n_alignments_from_bowtie on line 1086 which is called on line 1058 by do_mapping_bowtie which is called by do_mapping which is called by do_iteration which is called by do_iterations which is called on line 1568 near the end of main().

In summary, EM does not initialize any n_alignments variable but immediately tries to process the BAM file which attempts to use the n_alignments variable as part of a numpy array. Since n_alignments doesn't exist, the numpy array cannot initialize and the program crashes, i.e., the function ultimately requiring n_alignments is called before the function generating n_alignments. It seems that if no BAM file is provided to EMIRGE then it will derive n_alignments from bowtie's stderr but if you provide a BAM file, there's no way to provide this info and the program crashes.

**Comment by epruesse** [comment]
Alex,

thanks for reporting this. I've pushed a patch that should fix the issue.

Elmar
